yjit/src/core.rs

```diff
@@ -526,111 +535,238 @@ impl BranchTarget {
+            BranchTarget::Block(blockref) => Some(unsafe { blockref.as_ref() }.start_addr),
         }
     }
 
     fn get_blockid(&self) -> BlockId {
         match self {
-            BranchTarget::Stub(stub) => stub.id,
-            BranchTarget::Block(blockref) => blockref.borrow().blockid,
+            BranchTarget::Stub(stub) => BlockId { iseq: stub.iseq.get(), idx: stub.iseq_idx },
+            BranchTarget::Block(blockref) => unsafe { blockref.as_ref() }.get_blockid(),
         }
     }
 
     fn get_ctx(&self) -> Context {
         match self {
             BranchTarget::Stub(stub) => stub.ctx.clone(),
-            BranchTarget::Block(blockref) => blockref.borrow().ctx.clone(),
+            BranchTarget::Block(blockref) => unsafe { blockref.as_ref() }.ctx.clone(),
         }
     }
 
     fn get_block(&self) -> Option<BlockRef> {
         match self {
             BranchTarget::Stub(_) => None,
-            BranchTarget::Block(blockref) => Some(blockref.clone()),
+            BranchTarget::Block(blockref) => Some(*blockref),
         }
     }
 
-    fn set_iseq(&mut self, iseq: IseqPtr) {
+    fn set_iseq(&self, iseq: IseqPtr) {
         match self {
-            BranchTarget::Stub(stub) => stub.id.iseq = iseq,
-            BranchTarget::Block(blockref) => blockref.borrow_mut().blockid.iseq = iseq,
+            BranchTarget::Stub(stub) => stub.iseq.set(iseq),
+            BranchTarget::Block(blockref) => unsafe { blockref.as_ref() }.iseq.set(iseq),
         }
     }
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 struct BranchStub {
     address: Option<CodePtr>,
-    id: BlockId,
+    iseq: Cell<IseqPtr>,
+    iseq_idx: IseqIdx,
     ctx: Context,
 }
 
 /// Store info about an outgoing branch in a code segment
 /// Note: care must be taken to minimize the size of branch objects
-#[derive(Debug)]
 pub struct Branch {
     // Block this is attached to
     block: BlockRef,
 
     // Positions where the generated code starts and ends
-    start_addr: Option<CodePtr>,
-    end_addr: Option<CodePtr>, // exclusive
+    start_addr: CodePtr,
+    end_addr: Cell<CodePtr>, // exclusive
 
     // Branch target blocks and their contexts
-    targets: [Option<Box<BranchTarget>>; 2],
+    targets: [Cell<Option<Box<BranchTarget>>>; 2],
 
     // Branch code generation function
     gen_fn: BranchGenFn,
 }
 
+/// A [Branch] for a [Block] that is under construction.
+/// Fields correspond, but may be `None` during construction.
+pub struct PendingBranch {
+    /// Allocation holder for the address of the constructed branch
+    /// in error paths Box deallocates it.
+    uninit_branch: Box<MaybeUninit<Branch>>,
+
+    /// Branch code generation function
+    gen_fn: BranchGenFn,
+
+    /// Positions where the generated code starts and ends
+    start_addr: Cell<Option<CodePtr>>,
+    end_addr: Cell<Option<CodePtr>>, // exclusive
+
+    /// Branch target blocks and their contexts
+    targets: [Cell<Option<Box<BranchTarget>>>; 2],
```

Comment:
```suggestion
    /// Positions where the generated code starts and ends
    start_addr: Cell<Option<CodePtr>>,
    end_addr: Cell<Option<CodePtr>>, // exclusive

    /// Branch target blocks and their contexts
    targets: [Cell<Option<Box<BranchTarget>>>; 2],
    
    /// Branch code generation function
    gen_fn: BranchGenFn,
```

It might be easier to compare and understand the difference if the remaining fields are ordered in the same way.